### DISCUSSION TITLE
How to compare an array column and differentiate between an exact match and non-empty intersection?

### DISCUSSION THREAD

**Discussion by adelinor** [moj-analytical-services/splink#1053] [Q&A] [answered]
Hi,

I would like to write a comparison for an array column with levels:
- Null level if one of two arrays is empty
- exact match between the two arrays: all values in left are in right, and all values in right are in left
- at least one value in common
- 0 for anything else

In the example below, the comparison scores calculated with `array_intersect_at_sizes` are:
Codes comparison for (100,101)=1
Codes comparison for (100,102)=2
Codes comparison for (100,103)=0
Codes comparison for (100,104)=1
Codes comparison for (101,102)=1
Codes comparison for (101,103)=0
Codes comparison for (101,104)=1
Codes comparison for (102,103)=0
Codes comparison for (102,104)=1
Codes comparison for (103,104)=0

With this approach, a complete match for small arrays, pair (101, 104) in the example, does not get the highest score. Would it be feasible to give the highest score to an exact match?

```python
import pandas as pd

from splink.duckdb.duckdb_linker import DuckDBLinker
import splink.duckdb.duckdb_comparison_library as cl

df = pd.DataFrame({
        'unique_id': [100,         101,    102,          103,    104],
        'codes': [   ['ABC','DEF'],['ABC'],['DEF','ABC'],['GHI'],['ABC']]
    })

settings = {
            "link_type": "dedupe_only",
           "comparisons": [
                cl.array_intersect_at_sizes("codes", [2, 1]),
            ],
}

linker = DuckDBLinker(df, settings)
df_e = linker.predict().as_pandas_dataframe()

for i in range(100,104):
    for j in range(i + 1, 105):
        row = dict(
                df_e.query(
                    "unique_id_l == {} and unique_id_r == {}".format(i, j)
                ).iloc[0]
        )
        print("Codes comparison for ({},{})={}".format(i,j,row["gamma_codes"]))
```

Kind regards

**Comment by NickCrews** [comment, marked as answer]
Hi!

You are going to have to write some custom SQL for this. I haven't tested this, but something like:

```python
# Depending on which version of splink you're on, this typo of duckdb/duckb may be present.
from splink.duckdb import duckb_base as duckdb_base
# from splink.duckdb import duckdb_base as duckdb_base
from splink.duckdb import duckdb_comparison_level_library as cll

n_in_common = duckdb_base.size_array_intersect_sql("codes_l", "codes_r")

comparisons = [
    {
        "label_for_charts": "at least one is empty",
        "sql_condition": "length(codes_l) = 0 OR length(codes_r) = 0",
        "is_null_level": True,
    },
    {
        "label_for_charts": "100% match",
        "sql_condition": f"length(codes_l) = length(codes_r) AND length(codes_l) > 0 AND {n_in_common} = length(l.codes)",  # noqa: E501
    },
    {
        "label_for_charts": "At least element in common",
        "sql_condition": f"{n_in_common} > 0",
    },
    cll.else_level(),
]
```

**Comment by adelinor** [comment]
Hi @NickCrews ,

Your suggestion works exactly as I wanted to.

For the record, it might help others to mention that `comparisons` array has to be supplied to the `"comparison_levels"` attribute of a custom comparator definition. Also there is a small mistypo at the end 2nd SQL condition: should be codes_l.

A great thank you.

**Comment by Yash00git** [comment]
Hi, 

I have 2 arrays with different length. I am trying to use array_intersect_at_sizes but it takes 2 parameters. Can you please suggest, what values do I need to pass?

**Reply by RobinL** [reply]
Docs here:
https://github.com/moj-analytical-services/splink/blob/3fa517f1958dfc4873b240009429f2dd8c5cce5a/splink/comparison_library.py#L901

**Reply by Yash00git** [reply]
Hi Robin,

Thanks for sharing the link. I have gone through the documentation, but I have query regarding the below syntax.

'''
A comparison of the data in array column col_name with various intersection sizes to assess similarity levels.
An example of the output with default arguments and setting size_or_sizes = [3, 1] would be
Intersection has at least 3 elements
Intersection has at least 1 element (i.e. 1 or 2)
Anything else (i.e. empty intersection)
'''

What does this exactly mean by intersection of elements with 'size_or_sizes = [3, 1]'?
